opt-dist: rebuild rustc when doing static LLVM builds

ognevny opened the pull request:
when building LLVM it's obvious that in case of shared build rustc doesn't need to be recompiled, but with static builds it would be better to compile rustc again to ensure we linked proper library.
maybe I didn't understand the pipeline correctly, but it was strange for me to see that in Stage 5 LLVM is built while rustc is not

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

r? @Kobzol

try-job: dist-x86_64-msvc
try-job: dist-x86_64-linux

ognevny commented:
reverted stage 2 change as comment is pretty self-explanatory

Kobzol commented:
@bors2 try

> but it was strange for me to see that in Stage 5 LLVM is built while rustc is not

In stage 5 we shouldn't be building neither rustc nor LLVM. Bootstrap already skips building LLVM if it exists on disk, which is why we don't need to pass anything like `avoid_llvm_rebuild`.

ognevny commented:
> In stage 5 we shouldn't be building neither rustc nor LLVM. Bootstrap already skips building LLVM if it exists on disk, which is why we don't need to pass anything like `avoid_llvm_rebuild`.

this happens only if BOLT stage is run. otherwise right after llvm cleanup at stage 2 stage 5 starts

ognevny commented:
you can refer to [this](https://github.com/msys2/MINGW-packages/actions/runs/16205947685/job/45756147569) GHA job to see how the pipeline works on MSYS2 environment

ognevny commented:
> Could you please post some comparison benchmarks with/without this change?

sorry, I had literally _no skill_ at good benchmarking, and I need to ensure that I have "clean" environment on my machine...

if no one have time, I can try to do so tomorrow by the evening, I guess

ognevny commented:
thanks for the hint! will do soon

Kobzol commented:
I wanted to get more hands-on with Rust on Windows for unrelated reasons, so I rebooted into Windows and tried it there (a horrible experience, tbh :D). I found a 10% performance difference.. but in the wrong direction, lol.

```
D:\pg\rust\serde>hyperfine "cargo +56835d7ac14da9f966e1ff39fd9ffd2e29b764d1 build --release" "cargo +666996a7659a64c146e4765a1bba6ddc563819d6 build --release" --prepare "rmdir /s /q target"
Benchmark 1: cargo +56835d7ac14da9f966e1ff39fd9ffd2e29b764d1 build --release
  Time (mean ± σ):      9.466 s ±  0.483 s    [User: 87.788 s, System: 2.484 s]
  Range (min … max):    8.749 s … 10.174 s    10 runs

Benchmark 2: cargo +666996a7659a64c146e4765a1bba6ddc563819d6 build --release
  Time (mean ± σ):     10.369 s ±  0.216 s    [User: 98.415 s, System: 2.568 s]
  Range (min … max):   10.111 s … 10.748 s    10 runs

Summary
  cargo +56835d7ac14da9f966e1ff39fd9ffd2e29b764d1 build --release ran
    1.10 ± 0.06 times faster than cargo +666996a7659a64c146e4765a1bba6ddc563819d6 build --release
```

Let's see if you can reproduce.. Btw I confirmed that stage2 rustc was now rebuilt on MSVC during the final stage (it wasn't before).

ognevny commented:
> You can install these commits using https://github.com/kennytm/rustup-toolchain-install-master.

I can't download cargo with it: 
```
rustup-toolchain-install-master 56835d7ac14da9f966e1ff39fd9ffd2e29b764d1 --component=cargo -i x86_64-pc-windows-msvc
toolchain `56835d7ac14da9f966e1ff39fd9ffd2e29b764d1` is already installed
```

Kobzol commented:
I don't think you need cargo, you can just use whatever you have locally. But if you want it, delete the toolchain first (`rustup toolchain remove 56835d7ac14da9f966e1ff39fd9ffd2e29b764d1`), and then install it again.

ognevny commented:
```console
$ hyperfine "cargo +56835d7ac14da9f966e1ff39fd9ffd2e29b764d1 build --release" "cargo +666996a7659a64c146e4765a1bba6ddc563819d6 build --release" --prepare "cargo clean" --runs 5
Benchmark 1: cargo +56835d7ac14da9f966e1ff39fd9ffd2e29b764d1 build --release
  Time (mean ± σ):     10.883 s ±  0.491 s    [User: 77.163 s, System: 5.557 s]
  Range (min … max):   10.260 s … 11.633 s    5 runs

Benchmark 2: cargo +666996a7659a64c146e4765a1bba6ddc563819d6 build --release
  Time (mean ± σ):     11.376 s ±  0.291 s    [User: 82.654 s, System: 5.692 s]
  Range (min … max):   11.085 s … 11.855 s    5 runs

Summary
  cargo +56835d7ac14da9f966e1ff39fd9ffd2e29b764d1 build --release ran
    1.05 ± 0.05 times faster than cargo +666996a7659a64c146e4765a1bba6ddc563819d6 build --release
```

Kobzol commented:
Hmm, kinda hard to say what happens there. Maybe completely avoid skipping anything, and just build the whole of rustc and libstd from scratch, and we can try again.

Kobzol commented:
Yeah, and maybe also `--keep-stage`.

ognevny commented:
> Yeah, and maybe also `--keep-stage`.

done

Kobzol commented:
@bors2 try

bors commented:
:umbrella: The latest upstream changes (presumably #144145) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

mati865 commented:
Amusingly this seems to work with regular windows-gnu targets: https://github.com/msys2/MINGW-packages/pull/22778#issuecomment-2817305186

Yeah, Windows is just a mess.

ognevny commented:
> Yeah, and maybe also `--keep-stage`.

btw it's probably better to leave `--keep-stage 0` at least. otherwise our pipeline in MSYS2 may timeout

ognevny committed:
opt-dist: rebuild rustc when doing static LLVM builds

when building LLVM it's obvious that in case of shared build rustc doesn't need to be recompiled,
but with static builds it would be better to compile rustc again to ensure we linked proper library.
maybe I didn't understand the pipeline correctly, but it was strange for me to see that in Stage 5
LLVM is built while rustc is not

ognevny commented:
@Kobzol will we retry after LLVM 21 update?

Kobzol commented:
Yeah, why not.

ognevny commented:
@Kobzol can you do a try job here and for #144255, please?

ognevny commented:
huge regression!
```console
$ hyperfine "cargo +cd434309efcf5dc68b253e5ef6ba40c1c43711c9 build --release" "cargo +9c5e6d0dae78625bef89f77833e1
e4af5825bd24 build --release" --prepare "cargo clean" --runs 5
Benchmark 1: cargo +cd434309efcf5dc68b253e5ef6ba40c1c43711c9 build --release
  Time (mean ± σ):      9.772 s ±  0.114 s    [User: 73.706 s, System: 4.986 s]
  Range (min … max):    9.584 s …  9.887 s    5 runs

Benchmark 2: cargo +9c5e6d0dae78625bef89f77833e1e4af5825bd24 build --release
  Time (mean ± σ):     10.335 s ±  0.259 s    [User: 78.559 s, System: 5.008 s]
  Range (min … max):   10.169 s … 10.782 s    5 runs

Summary
  cargo +cd434309efcf5dc68b253e5ef6ba40c1c43711c9 build --release ran
    1.06 ± 0.03 times faster than cargo +9c5e6d0dae78625bef89f77833e1e4af5825bd24 build --release
```

Kobzol commented:
Hmm, I guess we should merge both PRs. Kinda sad about the Windows PGO situation.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-timer commented:
Queued 9171ddf5421f4c2c5693cb6368a34247697b325c with parent d20509c2a0c71b60aa2b51566e4d14920e8a1661, future [comparison URL](https://perf.rust-lang.org/compare.html?start=d20509c2a0c71b60aa2b51566e4d14920e8a1661&end=9171ddf5421f4c2c5693cb6368a34247697b325c).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.0 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

jhpratt commented:
still in queue

@bors r-

ognevny commented:
@rustbot modify labels: -S-waiting-on-author +merged-by-bors

ognevny commented:
it could at least stop waiting for me... but ok

jhpratt commented:
I wouldn't worry about labels on a closed/merged PR, but I've changed them regardless.
